File: ioi2020-biscuits/solution/biscuits-yanhao-smallAns.java
class biscuits {
	long[] arr = new long[(int) 2e5+5];

	long count_tastiness(long x, long[] a) {
		long[] s = new long[61];
		s[0] = a[0];
		for(int i=1; i<a.length; i++) {
			s[i] = s[i-1] + (a[i]<<i);
		}
		for(int i=a.length; i<=60; i++) {
			s[i] = s[i-1];
		}
		arr[0] = 0;
		int sz = 1;
		for(int i=0; i<s.length; i++) {
			long threshold = Math.min(1L<<i, s[i]/x-(1L<<i)+1);
			for(int j=0; arr[j]<threshold; j++) {
				arr[sz++] = arr[j]+(1L<<i);
			}
		}
		return sz;
	}
}

File: ioi2020-biscuits/solution/biscuits-yanhao-smallAns.cpp
#include "biscuits.h"
#include <vector>
#include <cstdio>
#include <assert.h>
using namespace std;

long long arr[(int) 4e7];
long long count_tastiness(long long x, std::vector<long long> a) {
    while(a.size()<=60) {
        a.push_back(0);
    }
    vector<long long> s;
    s.push_back(a[0]);
    for(int i=1; i<(int) a.size(); i++) {
        s.push_back(s.back() + (a[i]<<i));
    }
    arr[0] = 0;
    int sz = 1;
    for(int i=0; i<(int) s.size(); i++) {
        long long threshold = min(1LL<<i, s[i]/x-(1LL<<i)+1);
        for(int j=0; arr[j]<threshold; j++) {
            arr[sz++] = arr[j]+(1LL<<i);
        }
    }
    /*vector<long long> ans;
    ans.reserve((int) 4e7);
    ans.push_back(0);
    for(int i=0; i<(int) s.size(); i++) {
        if((x<<i)>(1LL<<60)) break;
        long long threshold = min(1LL<<i, s[i]/x-(1LL<<i)+1);
        for(int j=0; ans[j]<threshold; j++) {
            ans.push_back(ans[j]+(1LL<<i));
        }
    }*/

    //return ans.size();
    return sz;
}
